service-registry/registry-service-center/src/main/java/org/apache/servicecomb/serviceregistry/auth/TokenCacheManager.java

```diff
@@ -0,0 +1,191 @@
+
+    private final String accountName;
+
+    private final String password;
+
+    private final Clock clock;
+
+    private String token;
+
+    private long nextRefreshTime;
+
+    private boolean wrongPassword;
+
+    /**
+     * The life cycle period of a token, in millisecond.
+     * After the {@code tokenLife} time since the token created, it should be refreshed.
+     * <p>
+     * Default life time in sc is 30min, give 2min buffer
+     * </p>
+     */
+    private long tokenLife = TimeUnit.MINUTES.toMillis(30 - 2);
+
+    private ScheduledExecutorService tokenCacheWorker;
+
+    private Cipher cipher;
+
+    public TokenCache(String registryName, String accountName, String password,
+        Cipher cipher, Clock clock) {
+      this.registryName = registryName;
+      this.accountName = accountName;
+      this.password = password;
+      this.cipher = cipher;
+      this.clock = clock;
+    }
+
+    public String getToken() {
+      return token == null ? "" : token;
+    }
+
+    public void setTokenCacheWorker(ScheduledExecutorService tokenCacheWorker) {
+      Objects.requireNonNull(tokenCacheWorker, "input tokenCacheWorker is null");
+      if (this.tokenCacheWorker != null) {
+        throw new IllegalStateException("tokenCacheWorker already set!");
+      }
+
+      this.tokenCacheWorker = tokenCacheWorker;
+      startTokenRefreshTask();
+    }
+
+    private void startTokenRefreshTask() {
+      this.tokenCacheWorker.scheduleAtFixedRate(
+          new SuppressedRunnableWrapper(() -> {
+            if (isTokenOutdated()) {
+              refreshToken();
+            }
+          }),
+          1,
+          5,
+          TimeUnit.SECONDS);
+    }
+
+    private boolean isTokenOutdated() {
+      return clock.millis() > nextRefreshTime;
+    }
+
+    private void refreshToken() {
+      if (wrongPassword) {
+        return;
+      }
+      ServiceRegistry serviceRegistry = RegistryUtils.getServiceRegistry(registryName);
+      ServiceRegistryClient serviceRegistryClient =
+          serviceRegistry == null ? null : serviceRegistry.getServiceRegistryClient();
+      if ((serviceRegistry == null || serviceRegistryClient == null)
+          && ServiceRegistry.DEFAULT_REGISTRY_NAME.equals(registryName)) {
+        LOGGER.error("failed to get default serviceRegistry");
+        tokenCacheWorker.schedule( // retry after 1 second
+            this::refreshToken, 1, TimeUnit.SECONDS);
+        return;
+      }
+      RbacTokenRequest request = new RbacTokenRequest();
+      request.setAccountName(new String(cipher.decrypt(accountName.toCharArray())));
+      request.setPassword(new String(cipher.decrypt(password.toCharArray())));
+      RbacTokenResponse rbacTokenResponse = serviceRegistryClient.getRbacToken(request);
+      LOGGER.info("refresh token successfully {}", rbacTokenResponse.getStatusCode());
+      if (StringUtils.isEmpty(this.token) && Status.UNAUTHORIZED.getStatusCode() == rbacTokenResponse.getStatusCode()) {
```

Comment:
should not retry when using old version of service center

Reply:
done